contracts/AgentMech.sol

```diff
@@ -265,36 +340,39 @@ contract AgentMech is ERC721Mech {
         }
         _locked = 2;
 
-        // Get an account to deliver request to
-        address account = mapRequestAddresses[requestId];
-        // The account is zero if the delivery mech is different from a priority mech, or if request does not exist
-        if (account == address(0)) {
-            if (mechMarketplace != address(0)) {
-                account = IMechMarketplace(mechMarketplace).getMechDeliveryInfo(requestId).account;
-            }
+        // Check for the marketplace existence
+        if (mechMarketplace != address(0)) {
+            revert MarketplaceExists(mechMarketplace);
+        }
 
-            // Check if request exists in the mech marketplace
-            if (account == address(0)) {
-                revert RequestIdNotFound(requestId);
-            }
+        // Request delivery
+        _deliver(requestId, data);
 
-            // Increase the total number of requests, as the request is delivered by this mech
-            mapRequestsCounts[account]++;
-            numTotalRequests++;
-        } else {
-            // The account is non-zero if it is delivered by the priority mech
-            _cleanRequestInfo(account, requestId);
-        }
+        _locked = 1;
+    }
 
-        // Perform a pre-delivery of the data if it needs additional parsing
-        bytes memory requestData = _preDeliver(account, requestId, data);
+    /// @dev Delivers a request by a marketplace.
+    /// @notice This function ultimately calls mech marketplace contract to finalize the delivery.
+    /// @param requestId Request id.
+    /// @param data Self-descriptive opaque data-blob.
+    /// @param mechServiceId Mech operator service Id.
+    function deliverMarketplace(uint256 requestId, bytes memory data, uint256 mechServiceId) external onlyOperator {
```

Comment:
When delivering via a marketplace, the mech needs to provide its service Id in order to validate in the marketplace. Discussed with Ardian, he said it should be very easy to do. We could also figure out each specific mech's service Id, but it would take O(n) operations.